src/main/java/jcifs/netbios/NameServiceClientImpl.java

```diff
@@ -0,0 +1,1072 @@
+
+                }
+                catch ( InterruptedException ie ) {
+                    throw new InterruptedIOException();
+                }
+                finally {
+                    this.responseTable.remove(nid);
+                }
+
+                synchronized ( this.LOCK ) {
+                    if ( isWINS(request.addr) == false )
+                        break;
+                    /*
+                     * Message was sent to WINS but
+                     * failed to receive response.
+                     * Try a different WINS server.
+                     */
+                    if ( request.addr == getWINSAddress() )
+                        switchWINS();
+                    request.addr = getWINSAddress();
+                }
+            }
+        }
+    }
+
+
+    NbtAddress[] getAllByName ( Name name, InetAddress addr ) throws UnknownHostException {
+        int n;
+        Configuration config = this.transportContext.getConfig();
+        NameQueryRequest request = new NameQueryRequest(config, name);
+        NameQueryResponse response = new NameQueryResponse(config);
+
+        request.addr = addr != null ? addr : getWINSAddress();
+        request.isBroadcast = request.addr == null;
+
+        if ( request.isBroadcast ) {
+            request.addr = this.baddr;
+            n = config.getNetbiosRetryCount();
+        }
+        else {
+            request.isBroadcast = false;
+            n = 1;
+        }
+
+        do {
+            try {
+                send(request, response, config.getNetbiosRetryTimeout());
+            }
+            catch ( IOException ioe ) {
+                log.info("Failed to send nameservice request for " + name.name, ioe);
+                throw new UnknownHostException(name.name);
+            }
+
+            if ( response.received && response.resultCode == 0 ) {
+                return response.addrEntry;
+            }
+        }
+        while ( --n > 0 && request.isBroadcast );
+
+        throw new UnknownHostException(name.name);
+    }
+
+
+    NbtAddress getByName ( Name name, InetAddress addr ) throws UnknownHostException {
+        NameQueryRequest request = new NameQueryRequest(this.transportContext.getConfig(), name);
+        NameQueryResponse response = new NameQueryResponse(this.transportContext.getConfig());
+
+        if ( addr != null ) { /*
+                               * UniAddress calls always use this
+                               * because it specifies addr
+                               */
+            request.addr = addr; /* if addr ends with 255 flag it bcast */
+            request.isBroadcast = ( addr.getAddress()[ 3 ] == (byte) 0xFF );
+
+            int n = this.transportContext.getConfig().getNetbiosRetryCount();
+            do {
+                try {
+                    send(request, response, this.transportContext.getConfig().getNetbiosRetryTimeout());
+                }
+                catch ( InterruptedIOException ioe ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)